src/backend/services/mining-api/src/routes.py

```diff
@@ -54,20 +65,54 @@ def add_data_to_queue():
     return jsonify(success=True), HttpCode.OK.value
     
 
+def valid_proof(hash_, nonce):
+    
+    valid = (hash_.startswith('0' * difficulty))
+    return valid and hash_ == sha256(
+        (str(nonce) + ongoing_transaction["to"] + ongoing_transaction["from"] + str(ongoing_transaction["amount"]) + ongoing_transaction["id"] + ongoing_transaction["signature"]).encode('utf-8')
+      ).hexdigest();
+
+@socketio.on('connect')
+def client_connect():
+    global connected_clients
+    connected_clients += 1
+    print(connected_clients)
+
+@socketio.on('disconnect')
+def client_disconnect():
+    global connected_clients
+    connected_clients -= 1
+    print(connected_clients)
+
 @socketio.on('echo')
 def handle(message):
     socketio.emit("response", message)
 
 @socketio.on('proof')
 def handle_proofs(message):
-    transactions.ready_to_mine()
-    # TODO if proof for wrong id is received compare it with the current ongoing proof
-
-    # TODO send the transaction to blockchain
-
-    # TODO reward the miner (create a new transaction and send it to blockchain)
-
-    socketio.emit("stop_finding", message)
+    global ongoing_proof, ongoing_transaction, COINBASE_AMOUNT, blockchain_url, blockchain_wallet_url, transactions
+    
+    if (ongoing_proof == message['id']):
+        if valid_proof(message["proof"], message["nonce"]):
+
+            # ignore the rest of the clients that try to send the proof
+            ongoing_proof = None
+            ongoing_transaction = None
+            
+            socketio.emit("stopProof", {})
+            # new transactions are now ready to be mined    
+            transactions.ready_to_mine()
+
+            block_data = {}
+            # send transactions to blockchain
+            for key, value in message.items():
```

Comment:
Not a suggestion but interesting that the message from the socket comes in the form of a dict and not an object! Wouldn't have thought that

Reply:
Yeah looks like socketio handles JSON to dictionary conversion, which is pretty interesting.